### PR TITLE
fastp version

### PR DESCRIPTION
**Pull request by danrlu**
Features:
- Used fastp for trimming
- For raw files do: md5sum, fastqc and trim. 
- Combine pre-trim fastqc and trimming report by MultiQC. 
- In dsl2 format.

To be improved:
- The implementation of pre_trim_md5sum() is not very Nextflow-y. I couldn't find a better way.

Please see what changes need to be done. Thanks!

### CODE REVIEW

File: main.nf [hunk 1]
```diff
@@ -1,234 +1,66 @@
...
-    =======================
+#!/usr/bin/env nextflow 
+/*
+    Andersen Lab C. elegans Trimming Pipeline
+    Authors:
+    - Dan Lu <dan.lu@northwestern.edu>
 */
 
-process md5sum_pre {
+params.fastq_path="/projects/b1059/data/fastq/WI/dna/raw"
```
Comment (danielecook) [inline main.nf]: Don't include absolute paths in the `main.nf` file. Any absolute path specifications should go in the config files (e.g. `quest.conf`)

File: main.nf [hunk 2]
```diff
@@ -1,234 +1,172 @@
...
+params.trimmed_folder="${params.fastq_folder}_fastp"
+
+
 
 /* 
-    ======================
-    Perform initial FASTQC
-    ======================
+    ==================================================
+    get md5sum for raw files. do all file in 1 process
```
Comment (danielecook) [inline main.nf]: `Calculate MD5 for all files in a single process`

File: main.nf [hunk 3]
```diff
@@ -1,234 +1,172 @@
...
+
     output:
-        file("multiqc_report_pre.html")
+    /* note each item in the channel is a file pair */
+       path "*.html" // output to fastqc folder
+       path "*.zip", emit: pre_trim_fastqc_zip // emit for the next process
+
     """
-        multiqc --filename multiqc_report_pre.html ${params.directory}/fastqc
+    fastqc $fq1
```
Comment (danielecook) [inline main.nf]: Why are you running fastqc here? I thought you were moving to fastp...?

File: main.nf [hunk 4]
```diff
@@ -1,234 +1,172 @@
...
-    tag { dataset_id } 
-    
     input:
-        set dataset_id, file(forward), file(reverse) from trimmomatic_read_pairs
+      tuple sampleID, path(fq1), path(fq2) 
 
     output:
-        set dataset_id, file("${dataset_id}_1P.fq.gz"), file("${dataset_id}_2P.fq.gz") into trim_output
+      path "*_trimmed.fq.gz" 
+      path "*_fastp.json", emit: fastp_json
```
Comment (danielecook) [inline main.nf:31]: This file should have most/all metrics of interest for multiqc, so you should not need to run fastqc as you did before.

File: nextflow.config [hunk 5]
```diff
@@ -0,0 +1,11 @@
+process.cache = 'deep'
+
+
+report {
+    enabled = true
+}
+
+
+process {
+	conda = "/projects/b1059/software/conda_envs/trim-nf_env"
```
Comment (danielecook) [inline nextflow.config]: I would recommend bundling this with the workflow

File: main.nf [hunk 6]
```diff
@@ -1,234 +1,172 @@
...
-        outlog << "pyenv versions".execute().text
-        outlog << "--------ENV--------"
-        outlog << "ENV".execute().text
-        outlog << "--------brew--------"
-        outlog << "brew list".execute().text
-        outlog << "--------R--------"
-        outlog << "Rscript -e 'devtools::session_info()'".execute().text
-    }
+pre_trim_md5sum()  // check sum for all files. 
+// I don't know how to get the .fq out of fq channel (it has sampleID, fq1, fq2), so this process doesn't take input channel and run locally in the data folder.
```
Comment (danielecook) [inline main.nf]: I'll have a look later.

File: nextflow.config (hunk 5 above)
Comment (danielecook) [reply nextflow.config]: Also I would recommend using profiles.

File: nextflow.config [hunk 7]
```diff
@@ -0,0 +1,11 @@
+process.cache = 'deep'
```
Comment (danielecook) [inline nextflow.config]: I don't think you need to cache deep here. I might have added that..it's probably not necessary.

File: main.nf [hunk 8]
```diff
@@ -1,234 +1,66 @@
-#!/usr/bin/env nextflow
-
-params.email = 'dec@u.northwestern.edu'
-params.directory = "$PWD/"
-params.out = params.directory.replace("raw", "processed")
-println params.out
-params.threads = 8
-println "Running Trimmomatic on " + params.directory
```
Comment (danielecook) [inline main.nf:8]: These messages can be helpful when running

File: main.nf (hunk 1 above)
Comment (danielecook) [reply main.nf]: So you should be able to specify the path to raw data using the profile, and the folder name as an input parameter...

e.g.

`quest.conf` will have something like: 

```bash
params.fastq_folder="/projects/b1059/data/fastq/WI/dna/raw"
```
And when you run it something like this:

```
nextflow run main.nf --folder <folder name>
```

For testing you can supply both the fastq raw location and the folder name so you only have to specify:

```bash
nextflow run main.nf -profile debug
```

File: main.nf [hunk 9]
```diff
@@ -1,234 +1,66 @@
...
-    output:
-        file("md5sum.txt")
 
-    """
-        echo "Great!"
-    """
+process fastp_trim {
 
-}
+    conda "fastp"
```
Comment (danielecook) [inline main.nf]: You should specify a specific version of fastp to use here.

File: main.nf [hunk 10]
```diff
@@ -1,234 +1,66 @@
...
 
-/* 
-    ================
-    Perform trimming
-    ================
+    publishDir "${params.trimmed_path}/${params.trimmed_folder}", mode: 'copy', pattern: "*.fq.gz"
+/*    publishDir "${params.trimmed_path}/${params.trimmed_folder}", mode: 'copy', pattern: "*_fastp.json"
```
Comment (danielecook) [inline main.nf]: Don't leave commented out code hanging around. You can always go back using git.

File: main.nf [hunk 11]
```diff
@@ -1,234 +1,66 @@
...
 }
 
 
-workflow.onComplete {
 
-    user="whoami".execute().text
+process multi_QC {
 
-    summary = """
+    conda "multiqc"
```
Comment (danielecook) [inline main.nf]: add the version here

**Review by danielecook — CHANGES_REQUESTED** [review]

File: main.nf (hunk 3 above)
Comment (danrlu) [reply main.nf]: I thought fastp doesn't run QC on pre-trimmed data. Now I see it does. I'll remove this part~

Comment (danrlu) [reply main.nf]: `b31a08f058e14fe4264b0e256f934792917fdfac`

File: main.nf (hunk 4 above)
Comment (danrlu) [reply main.nf:31]: `b31a08f058e14fe4264b0e256f934792917fdfac`

File: main.nf (hunk 1 above)
Comment (danrlu) [reply main.nf]: `06d6228`

File: main.nf (hunk 2 above)
Comment (danrlu) [reply main.nf]: `cd191f6`

File: nextflow.config (hunk 5 above)
Comment (danrlu) [reply nextflow.config]: `06d6228`

File: nextflow.config (hunk 7 above)
Comment (danrlu) [reply nextflow.config]: `06d6228`

File: main.nf (hunk 8 above)
Comment (danrlu) [reply main.nf:8]: `96d4a25`

File: main.nf (hunk 9 above)
Comment (danrlu) [reply main.nf]: I'll make a record of conda env list

File: main.nf (hunk 10 above)
Comment (danrlu) [reply main.nf]: oh no cannot find the commit hash! but I fixed this somewhere.

File: main.nf (hunk 11 above)
Comment (danrlu) [reply main.nf]: will be on the conda env list

File: main.nf (hunk 1 above)
Comment (danrlu) [reply main.nf]: in `d5f7a4c` changed profile quest to be named as 'standard' which nextflow uses by default if no profile is specified. So users can run it with `nextflow run main.nf` without specifying profile on Quest, which is likely the only place this pipeline is going to be used.

**Review by danielecook — APPROVED** [review]

**Pull request merged**